homeassistant/components/screenlogic/number.py

```diff
@@ -29,66 +86,113 @@ async def async_setup_entry(
+
+
+@dataclass
+class ScreenLogicNumberDescription(
+    NumberEntityDescription,
+    ScreenLogicEntityDescription,
+    ScreenLogicNumberRequiredMixin,
+):
+    """Describes a ScreenLogic number entity."""
+
 
 class ScreenLogicNumber(ScreenlogicEntity, NumberEntity):
-    """Class to represent a ScreenLogic Number."""
+    """Class to represent a ScreenLogic Number entity."""
 
-    _attr_has_entity_name = True
+    entity_description: ScreenLogicNumberDescription
 
-    def __init__(self, coordinator, data_key, enabled=True):
-        """Initialize of the entity."""
-        super().__init__(coordinator, data_key, enabled)
-        self._body_type = SUPPORTED_SCG_NUMBERS.index(self._data_key)
-        self._attr_native_max_value = SCG.LIMIT_FOR_BODY[self._body_type]
-        self._attr_name = self.sensor["name"]
-        self._attr_native_unit_of_measurement = self.sensor["unit"]
-        self._attr_entity_category = EntityCategory.CONFIG
+    def __init__(
+        self,
+        coordinator: ScreenlogicDataUpdateCoordinator,
+        entity_description: ScreenLogicNumberDescription,
+    ) -> None:
+        """Initialize a ScreenLogic number entity."""
+        self._set_value_func = entity_description.set_value
+        self._set_value_params = entity_description.set_value_params
+        super().__init__(coordinator, entity_description)
 
     @property
     def native_value(self) -> float:
         """Return the current value."""
-        return self.sensor["value"]
+        return self.entity_data[ATTR.VALUE]
 
     async def async_set_native_value(self, value: float) -> None:
         """Update the current value."""
-        # Need to set both levels at the same time, so we gather
-        # both existing level values and override the one that changed.
-        levels = {}
-        for level in SUPPORTED_SCG_NUMBERS:
-            levels[level] = self.gateway_data[SL_DATA.KEY_SCG][level]["value"]
-        levels[self._data_key] = int(value)
-
-        if await self.coordinator.gateway.async_set_scg_config(
-            levels[SUPPORTED_SCG_NUMBERS[BODY_TYPE.POOL]],
-            levels[SUPPORTED_SCG_NUMBERS[BODY_TYPE.SPA]],
-        ):
-            _LOGGER.debug(
-                "Set SCG to %i, %i",
-                levels[SUPPORTED_SCG_NUMBERS[BODY_TYPE.POOL]],
-                levels[SUPPORTED_SCG_NUMBERS[BODY_TYPE.SPA]],
+
+        # Current API requires certain values to be set at the same time. This
+        # gathers the existing values and updates the particular value being
+        # set by this entity.
+        args = {}
+        for data_path in self._set_value_params:
+            data_path = realize_path_template(data_path, self._data_path)
+            data_value = data_path[-1]
+            args[data_value] = self.coordinator.gateway.get_value(
+                *data_path, strict=True
             )
+
+        args[self._data_key] = value
+
+        if self._set_value_func(*args.values()):
```

Comment:
For a future PR (lets not make this bigger)

https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=qual#silver-
Operations like service calls and entity methods (e.g. Set HVAC Mode) have proper exception handling. Raise ValueError on invalid user input and raise HomeAssistantError for other failures such as a problem communicating with a device.